src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,627 @@
+
+// @defgroup sir_proto SIR protobuf
+// @brief This is a SIR description for Google's protocol buffers library
+// <a href="https://developers.google.com/protocol-buffers/">Protobuf</a>.
+
+syntax = "proto3";
+
+package dawn.sir.proto;
+
+// @brief Source information
+//
+// `(-1,-1)` indicates an invalid location.
+// @ingroup sir_proto
+message SourceLocation {
+  int32 Line = 1;   // Source line
+  int32 Column = 2; // Column in the `Line`
+}
+
+// @brief Stencil attributes
+//
+// Stencil attributes can be used to adjust optimization and code generation for individual stencil
+// and stencil_functions.
+// @ingroup sir_proto
+message Attributes {
+  bool no_codegen = 1;        // Don't generate code for this stencil
+  bool merge_stages = 2;      // Merge the Stages of this stencil
+  bool merge_do_methods = 3;  // Merge the Do-Methods of this stencil
+  bool merge_temporaries = 4; // Merge the temporaries of this stencil
+  bool use_k_caches = 5;      // Use K-Caches
+}
+
+// @brief Description of a field argument of a Stencil or StencilFunction
+// @ingroup sir_proto
+message Field {
+  string name = 1;        // Name of the field
+  SourceLocation loc = 2; // Source location
+  bool is_temporary = 3;  // Is the field a temporary?
+}
+
+// @brief Directional argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Direction {
+  string name = 1;        // Name of the directional argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Offset argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Offset {
+  string name = 1;        // Name of the offset argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Directional argument of a StencilFunction
+// @ingroup sir_proto
+message StencilFunctionArg {
+  // One of Field, Direction or Offset
+  oneof Arg {
+    Field field_value = 1;
+    Direction direction_value = 2;
+    Offset offset_value = 3;
+  }
+}
+
+// @brief Representation of a vertical interval, given by a lower and upper bound where a bound
+// is represented by a level and an offset (`bound = level + offset`)
+//
+// The Interval has to satisfy the following invariants:
+//  - `lower_level >= Interval::Start`
+//  - `upper_level <= Interval::End`
+//  - `(lower_level + lower_offset) <= (upper_level + upper_offset)`
+//
+// @ingroup sir_proto
+message Interval {
+  // Indication of a special level
+  enum SpecialLevel {
+    Start = 0; // The first level
+    End = 1;   // The last level
+  }
+
+  // Lower level
+  //
+  // A level can be any integer between `[Start, End]` while the `Start` and `End` levels are
+  // treated specially.
+  oneof LowerLevel {
+    SpecialLevel special_lower_level = 1;
+    int32 lower_level = 2;
```

Comment:
Maybe would be good to have a Level message, instead of simply an integer. 
```
message Level {
  string name = 0;
  int32 position = 1;
}
```
Like that an interval would always refer to an already defined level, rather an unsafe integer value. 
Also I think we need a way to identify if the level will be provided at compile time (optionally) or runtime. 
If position would be = -1, then the compiler would need to generate an API where the values of all levels are passed by runtime arguments

Reply:
Agreed, this requires lots of changes to the SIR and I suggest we tackle this problem in a separate PR as this one is getting quite big already.